// typescript-json-decoder/dist/primitive-decoders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.date = exports.nil = exports.undef = exports.boolean = exports.number = exports.string = void 0;
var pojo_1 = require("./pojo");
var string = function (s) {
    pojo_1.assert_is_pojo(s);
    if (typeof s !== 'string') {
        throw "The value `" + JSON.stringify(s) + "` is not of type `string`, but is of type `" + typeof s + "`";
    }
    return s;
};
exports.string = string;
var number = function (n) {
    pojo_1.assert_is_pojo(n);
    if (typeof n !== 'number') {
        throw "The value `" + JSON.stringify(n) + "` is not of type `number`, but is of type `" + typeof n + "`";
    }
    return n;
};
exports.number = number;
var boolean = function (b) {
    pojo_1.assert_is_pojo(b);
    if (typeof b !== 'boolean') {
        throw "The value `" + JSON.stringify(b) + "` is not of type `boolean`, but is of type `" + typeof b + "`";
    }
    return b;
};
exports.boolean = boolean;
exports.undef = (function (u) {
    pojo_1.assert_is_pojo(u);
    if (typeof u !== 'undefined') {
        throw "The value `" + JSON.stringify(u) + "` is not of type `undefined`, but is of type `" + typeof u + "`";
    }
    return u;
});
exports.nil = (function (u) {
    pojo_1.assert_is_pojo(u);
    if (u !== null) {
        throw "The value `" + JSON.stringify(u) + "` is not of type `null`, but is of type `" + typeof u + "`";
    }
    return u;
});
var date = function (value) {
    pojo_1.assert_is_pojo(value);
    var dateString = exports.string(value);
    var timeStampSinceEpoch = Date.parse(dateString);
    if (isNaN(timeStampSinceEpoch)) {
        throw "String `" + dateString + "` is not a valid date string";
    }
    return new Date(timeStampSinceEpoch);
};
exports.date = date;
